docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,182 @@
+
+### Declared Nullability
+
+Declared nullability is the nullability that was explicitly typed in the source code by the programmer. There are some cases where this is inferred, such as lambda arguments/return types, `var` variables, and type parameter substitution. For these cases, flow state is necessary to compute the declared nullability. For example:
+
+```C#
+string? s1 = string.Empty;
+var s2 = s1;
+s2 = null; // This will warn, because the type of s2 is string, not string?
+```
+
+The declared nullability of `s1` is `MaybeNull`, even though it is being initialized with a non null value. However, because of the flow state of `s1` is `NotNull`, we will infer the declared nullability of `s2` to be `NotNull` as well.
+
+Declared nullability can be retreived from the declaration symbol (e.g. `IFieldSymbol`, `IMethodSymbol`, etc.).
+
+### Flow State
+
+When the nullable feature is enabled, Roslyn will track the flow state of expressions throughout a method, regardless of what the variable was declared as. `SemanticModel` can be used to request flow state information for individual expressions or sub-expressions through the `GetTypeInfo` APIs. The `TypeInfo` struct will be augmented with `Nullability` and `ConvertedNullability` information, similar to how it currently has `Type` and `ConvertedType`. Nullability of symbols at a location can be retrieved through the `GetSymbolInfo` API: if the expression successfully resolved to a single symbol, the `Symbol` property will have nullability calculated. `CandidateSymbols` will not have calculated nullabilities, as we cannot run nullable analysis on code that has failed overload resolution.
+
+### `GetSingleStatementSemanticModel`
+
+This API proposal changes the way `SemanticModel` calls behave by default. Currently, if semantic information is requested for a `SyntaxNode`, we make an effort to bind as little as possible to provide semantic information about that statement, usually a single statement. However, if that semantic information will include `Nullability`, we must bind all statements that come before that statement and run nullability analysis up to and including that statement. We believe that this change will be ok for most scenarios, as things such as colorization and analyzers are already running in the background near-constantly, binding pretty much everything on screen and in the file anyway. There are a few specific scenarios that will be perf-sensitive, and `GetSingleStatementSemanticModel()` will be provided for these. It will share an initial-binding `BoundNode` cache with the standard semantic model, so it will get the benefits if something else with the same semantic model requests information, as it would have today.
```

Comment:
`GetSingleStatementSemanticModel()` seems a bit obscure as a name. At least, it seems scary for something that is effectively "give me a semantic model that does the same level of analysis as a normal semantic model used to". This strikes me as something that should be used in advanced scenarios, while the reality is that it is more perf-conscious.

Reply:
> I would expect users to see this name and assume that it's supposed to be used in advanced scenarios, while the reality is that it is more perf-conscious.

Well, in .NET it seems we have (and often still) treat performance as an advanced scenario, if that means less of fidelity or other shortcomings the consumer needs to be aware of. In your case, maybe you want to make it an overloads that takes parameter, like `GetSemanticModel(bool skipNullabilityInformation)`. The existing overload (logically) calls this with `false`.

Reply:
@terrajobst: I totally agree that we treat performance as an advanced scenario, and we should. I think this case is _slightly_ different since we're making the normal code path less efficient to provide more correctness and offering a way to get back the original behavior, which happens to be more efficient. I think you're suggested overload is a good one.

Reply:
> I would expect users to see this name and assume that it's supposed to be used in advanced scenarios, while the reality is that it is more perf-conscious.

Yes, and I this was my intent. I thought that in the design meeting we had come to the conclusion that there were really only a few IDE scenarios that we expected to need this perf saving, and that everyone else really won't get much of an advantage from this API.

The problem with `GetSemanticModel(bool skipNullabiliytInformation)` is twofold:
1. The nullability-unaware and nullability-aware semantic models will be unable to share a bound node cache. This _would_ simplify implementation, but one of the intents here was that the nullable-aware model would share the initial bind cache of bound nodes with the unaware model.
2. One of my intents with with API was to provide a way to get a performant `SemanticModel`, without saying precisely what we do to get that performance. This way, if we add future analysis passes to the default model, we don't overwhelm the user with options for turning off individual things.

Reply:
> The nullability-unaware and nullability-aware semantic models will be unable to share a bound node cache.

Why is that? If this is the right API shape, I would think we could adjust the implementation to make it work (`ConditionalWeakTable` may be your friend here :smile:).

> One of my intents with with API was to provide a way to get a performant SemanticModel, without saying precisely what we do to get that performance. This way, if we add future analysis passes to the default model, we don't overwhelm the user with options for turning off individual things.

I definitely agree with that principle, but the name `GetSingleStatementSemanticModel` is probably too specific to cover any possible future analysis passes as well. I think you could achieve your goal here by picking a different name for the flag (e.g. `performFullAnalysis` with a default of true).

Reply:
@jasonmal, what do you think of this proposal from an IDE usage perspective? Give Dustin's points I'd probably be fine with a `GetSemanticModel(bool performFullAnalysis)`, but I'd like to hear your thoughts on usage.

Reply:
@terrajobst while I can see some value in negating the boolean, I'm specifically avoiding the ability to configure multiple different analyses here. I don't think we intend this API for consumers to decide "oh, we're only going to need this part of the information so lets save on analysis time". My understanding of this API is that it's for _very_ specific scenarios that are extremely performance-constrained and only need information from initial binding.

Reply:
I endorse @terrajobst's proposal, if only because it's also what I was thinking in the meetings all along. I imagine one of two futures:

1. We never have a second type of analysis that we want to disable like this. Great, the scenario you are worried about won't happen.
2. We have a second type of analysis that you want to disable. Unless you're willing to bet me $1,000,000 (bitcoin not accepted) that it will fall in the same classification as this type _and_ all the features that wanted or didn't want it are the exact same set as today...you're going to have somebody who wants to turn one and one off. At this point your API won't work at all, and you'll be probably doing what @terrajobst is arguing for.

> One of my intents with with API was to provide a way to get a performant SemanticModel, without saying precisely what we do to get that performance.

SemanticModels are already magic, but fundamentally writing any analysis means you need to know what's happening. If you made SemanticModel faster by limiting it's analysis and returning a best guess after a few milliseconds of CPU...we'd be at your door with pitchforks because you just broke all of our stuff.

I assume your concern 1 about bound node caches is no longer a concern? If you could implement GetSingleStatementSemanticModel then I don't see how (today) the implementation of this other API would just be would be just:

`GetSemanticModel(bool skipNullability) => skipNullability ? GetSingleStatementSemanticModel() : this`

Which doesn't introduce new complexity.